provider/consul_catalog.go

```diff
@@ -90,21 +90,33 @@ func (provider *ConsulCatalog) healthyNodes(service string) (catalogUpdate, erro
 
 	set := map[string]bool{}
 	tags := []string{}
+	nodes := []*api.ServiceEntry{}
 	for _, node := range data {
-		for _, tag := range node.Service.Tags {
-			if _, ok := set[tag]; ok == false {
-				set[tag] = true
-				tags = append(tags, tag)
+		constraintTags := provider.getContraintTags(node.Service.Tags)
+		if ok, failingConstraint, err := provider.MatchConstraints(constraintTags); err != nil {
+			return catalogUpdate{}, err
+		} else if ok == true {
+			nodes = append(nodes, node)
+			// merge tags of every nodes in a single slice
+			// only if node match constraint
+			for _, tag := range node.Service.Tags {
+				if _, ok := set[tag]; ok == false {
+					set[tag] = true
+					tags = append(tags, tag)
+				}
 			}
+		} else if ok == false && failingConstraint != nil {
```

Comment:
If possible, and to stay homogeneous with Docker/Marathon backends, it would be great to filter services in a `serviecsFilter` method like in https://github.com/containous/traefik/blob/master/provider/marathon.go#L135 or in https://github.com/containous/traefik/blob/master/provider/docker.go#L173